ehsm_kms_service/quote_manager_apis.js

```diff
@@ -0,0 +1,270 @@
+ * @returns {boolean}
+ */
+function checkStringParam(param, required, maxLength) {
+    if (param == '' || param == undefined) {
+        if (required) {
+            return false
+        } else {
+            return true
+        }
+    } else {
+        if (typeof (param) == 'string') {
+            if (maxLength != undefined && param.length > maxLength) {
+                return false
+            } else {
+                return true
+            }
+        } else {
+            return false
+        }
+    }
+}
+
+/**
+ * check uuid format
+ * @param {String} uuid : a uuid string, eg. 0197ad2d-c4be-4948-996d-513c6f1e****
+ * @returns {boolean}
+ */
+function check_UUID_format(uuid) {
+    if (uuid != '' && uuid != undefined) {
+        if (!((/^([a-z\d]{8}-[a-z\d]{4}-[a-z\d]{4}-[a-z\d]{4}-[a-z\d]{12})$/).test(uuid))) {
+            return false
+        }
+    }
+    return true
+}
+
+/**
+ * Generate a quote of the eHSM-KMS core enclave
+ * @param {Object} res : response
+ * @param {Object} payload
+ *          ==> {r}challenge(String [1~1024]) : a string. eg. 'Y2hhbGxlbmdl'
+ * @returns {Object}
+ *          ==> {r}challenge(String) : a string. eg. 'Y2hhbGxlbmdl'
+ *          ==> {r}quote(String) : A quote for the eHSM-KMS core enclave format in BASE64 string. eg. 'AwACAAAAAAAHAAwAk5pB∗∗∗'
+ */
+const generateQuote = async (res, payload, action) => {
+    try {
+        const { challenge } = payload
+        if (checkStringParam(challenge, true, CHALLENGE_LENGTH_MAX)) {
+            napi_res = napi_result(action, res, [challenge])
+            napi_res && res.send(napi_res)
+        } else {
+            res.send(_result(400, `challenge cannot be empty, must be string and length not more than ${CHALLENGE_LENGTH_MAX}`))
+            return
+        }
+    } catch (e) {
+        console.info('generateQuote :: ', e)
+        res.send(_result(500, 'Server internal error, please contact the administrator.'))
+        return
+    }
+}
+
+/**
+ * Users are expected already got a valid DCAP format QUOTE. And it could use this API to send it to eHSM-KMS to do a quote verification.
+ * @param {Object} res : response
+ * @param {String} appid : appid of user
+ * @param {Object} DB : database controller
+ * @param {Object} payload
+ *          ==> {r}quote(String [1~10924]) : A valid DCAP quote in BASE64 string. eg. 'AwACAAAAAAAHAAwAk5pB∗∗∗'
+ *          ==> {r}nonce(String [1~1024]) : A nonce in BASE64 string. eg. 'bm9uY2U='
+ *          ==> {r}policyId(String [36]) : quote policy ID. eg. '326f2049-56ab-474f-a1c0-516de230****'
+ * @returns {Object}
+ *          ==> {r}result(bool) : The result of quote verification. eg. 'true'
+ *          ==> {r}nonce(bool) : The nonce in BASE64 string. eg. 'bm9uY2U=
+ *          ==> {r}sign(String) : The HAMC sign of result and nonce calculated by the API Key. eg. 'T4DRCEZAPLBbb+d3ObD∗∗∗'
+ */
```

Comment:
@syan10  Is there a requirement for the freshness of the verification result?  e.g. a nonce passed from the caller and will be calculated into the returned signature.

Reply:
nonce is currently provided in this API.